persist/fs/commitlog/reader.go

```diff
@@ -109,58 +165,171 @@ func (r *reader) Read() (
+}
 
-	metadata, ok := r.metadataLookup[entry.Index]
-	if !ok && len(entry.Metadata) != 0 {
-		r.metadataDecoder.Reset(entry.Metadata)
-		decoded, err := r.metadataDecoder.DecodeLogMetadata()
-		if err != nil {
-			resultErr = err
+func (r *reader) readLoop() {
+	index := 0
+	eofFound := false
+	for {
+		select {
+		case <-r.cancelCtx.Done():
+			for _, decoderBuf := range r.decoderBufs {
+				close(decoderBuf)
+			}
+			r.shutdownChan <- r.close()
 			return
+		default:
+			if eofFound {
+				continue
+			}
+			data, err := r.readChunk()
+			if err == io.EOF {
+				eofFound = true
+			}
+
+			// Distribute the decoding work in round-robin fashion so that when we
+			// read round-robin, we get the data back in the original order.
+			r.decoderBufs[index%r.numConc] <- decoderArg{
+				bytes: data,
+				err:   err,
+			}
+
+			index++
 		}
+	}
+}
 
-		id := r.bytesPool.Get(len(decoded.ID))
-		id.IncRef()
-		defer id.DecRef()
-		id.AppendAll(decoded.ID)
-
-		namespace := r.bytesPool.Get(len(decoded.Namespace))
-		namespace.IncRef()
-		defer namespace.DecRef()
-		namespace.AppendAll(decoded.Namespace)
-
-		metadata = Series{
-			UniqueIndex: entry.Index,
-			ID:          ts.BinaryID(id),
-			Namespace:   ts.BinaryID(namespace),
-			Shard:       decoded.Shard,
+func (r *reader) decoderLoop(inBuf <-chan decoderArg, outBuf chan<- readResponse) {
+	decodingOpts := r.opts.FilesystemOptions().DecodingOptions()
+	decoder := msgpack.NewDecoder(decodingOpts)
+	metadataDecoder := msgpack.NewDecoder(decodingOpts)
+
+	for arg := range inBuf {
+		readResponse := readResponse{}
+		if arg.err != nil {
+			readResponse.resultErr = arg.err
+			outBuf <- readResponse
+			continue
 		}
-		ok = true
 
-		r.metadataLookup[entry.Index] = metadata
-	}
+		arg.bytes.IncRef()
+		decoder.Reset(arg.bytes.Get())
+		entry, err := decoder.DecodeLogEntry()
+		if err != nil {
+			readResponse.resultErr = err
+			outBuf <- readResponse
+			continue
+		}
 
-	if !ok {
-		resultErr = errCommitLogReaderMissingLogMetadata
-		return
-	}
+		if len(entry.Metadata) != 0 {
+			metadataDecoder.Reset(entry.Metadata)
+			decoded, err := metadataDecoder.DecodeLogMetadata()
+			if err != nil {
+				readResponse.resultErr = err
+				outBuf <- readResponse
+				continue
+			}
 
-	series = metadata
-	datapoint = ts.Datapoint{
-		Timestamp: time.Unix(0, entry.Timestamp),
-		Value:     entry.Value,
+			id := r.bytesPool.Get(len(decoded.ID))
+			id.IncRef()
+			id.AppendAll(decoded.ID)
+
+			namespace := r.bytesPool.Get(len(decoded.Namespace))
+			namespace.IncRef()
+			namespace.AppendAll(decoded.Namespace)
+
+			r.metadata.Lock()
+			r.metadata.lookup[entry.Index] = Series{
+				UniqueIndex: entry.Index,
+				ID:          ts.BinaryID(id),
+				Namespace:   ts.BinaryID(namespace),
+				Shard:       decoded.Shard,
+			}
+			waitGroup, ok := r.metadata.wgs[entry.Index]
+			// Another goroutine is blocked waiting for this metadata
+			if ok {
+				waitGroup.Done()
```

Comment:
can have multiple results waiting for a metadata, a single Done won't do the trick. Maybe use a `sync.Cond` instead

Reply:
I didn't know about sync.Cond when I was writing this, but I think the existing solution is correct. If you look at it, the waitgroup will never increment higher than 1 so a single call to .Done() is enough, and you can have multiple goroutines waiting on the same waitgroup as seen here: https://play.golang.org/p/ivmvhwksFp

Reply:
yeah but if multiple go routines are waiting on the same waitgroup, none will be released with a single `Done()`.

Reply:
Are you sure? I tested it in the go playground and you have multiple goroutines waiting on the same waitgroup with a single call to .Done() and it works: https://play.golang.org/p/ivmvhwksFp

Reply:
only happens if you have more than 1 waiting go-routine, e.g. https://play.golang.org/p/mjcxBGv4Bn

Reply:
I'll add a comment explaining it

Reply:
Yeah but the difference is that in your example the waitgroup was incremented multiple times (I.E you called add twice). In the reader code I have, even if there are 4 goroutines waiting, Add() is only called once and they're all released with a single call to Done()